[SPIR-V] Fix bad insertion for type/id MIR

Keenuts opened the pull request:
Those instructions were inserted either after the instruction using it, or in the middle of the module.
The first directly causes an issue. The second causes a more subtle issue: the first type the type is inserted, the emission is fine, but the second times, the first instruction is reused, without checking its position in the function. This can lead to the second usage dominating the definition.

In SPIR-V, types are usually in the header, above all code definition, but at this stage I don't think we can, so what I do instead is to emit it in the first basic block.

This commit reduces the failed tests with expensive checks from 107 to 71.

Signed-off-by: Nathan Gauër <brioche@google.com>

Keenuts committed:
[SPIR-V] Fix bad insertion for type/id MIR

Those instructions were inserted either after the instruction
using it, or in the middle of the module.
The first directly causes an issue. The second causes a more subtle
issue: the first type the type is inserted, the emission is fine, but
the second times, the first instruction is reused, without checking its
position in the function. This can lead to the second usage dominating
the definition.

In SPIR-V, types are usually in the header, above all code definition,
but at this stage I don't think we can, so what I do instead is to emit it in
the first basic block.

This commit reduces the failed tests with expensive checks from 107 to
71.

Signed-off-by: Nathan Gauër <brioche@google.com>

another method

VyacheslavLevytskyy commented on llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.h:
```diff
@@ -336,6 +347,7 @@ class SPIRVGlobalRegistry {
   MachineFunction *setCurrentFunc(MachineFunction &MF) {
     MachineFunction *Ret = CurMF;
     CurMF = &MF;
+    LastInsertedType = nullptr;
```
The problem here would be when a user of this function switches contexts between machine functions. After this change the context would be not only `MachineFunction *` that is returned, but also last inserted type position. See, for example:

https://github.com/llvm/llvm-project/blob/372611e186fc7eaea4f94c3ee7516d5d0e655dcd/llvm/lib/Target/SPIRV/SPIRVISelLowering.cpp#L291-L295

Here the context is switched between function definition/function call, and we would reset `LastInsertedType` to nullptr in both to/fro cases.

Keenuts committed:
pr-feedback: keep state across context switch

The code can switch functions multiple times, and come
back to a previously visited function, meaning we need to keep
the last inserted type for each function.

Signed-off-by: Nathan Gauër <brioche@google.com>

VyacheslavLevytskyy commented:
cool fix overall!
